### ISSUE TITLE
Type-driven header lookup (for clients)

### ISSUE THREAD

**Issue by DanBurton** [haskell-servant/servant#1063]
Suppose in my Servant client I have a hold of some value with some headers.

```haskell
-- src/Lib.hs
{-# LANGUAGE DataKinds       #-}
{-# LANGUAGE TypeOperators   #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GADTs #-}
module Lib where

import Servant

newtype Blah = Blah String deriving (ToHttpApiData, FromHttpApiData)
type BlahHeader = "X-Blah-Header"

-- You can imagine that there are more headers than just the one
unitWithBlahHeader :: Headers '[Header BlahHeader Blah] ()
unitWithBlahHeader = addHeader @BlahHeader (Blah "blah-header-value") ()
```

Now suppose I want to look up what the `BlahHeader` is. I don't want to have to manually pattern match on the HList and know which exact position my desired header is in the list. The code I wish I could write instead looks like this:

```haskell
blahResponseHeader :: ResponseHeader BlahHeader Blah
blahResponseHeader = lookupResponseHeader @BlahHeader unitWithBlahHeader
```

However, I am not aware of any existing `lookupResponseHeader` function that behaves this way.

Here's a sample implementation of it:

```haskell
-- | Look up a specific ResponseHeader,
-- without having to know what position it is in the HList.
--
-- Property:
-- lookupResponseHeader @h (addHeader @h v a) == Header v
lookupResponseHeader :: (HasResponseHeader h a headers)
  => Headers headers r -> ResponseHeader h a
lookupResponseHeader = hlistLookupHeader . getHeadersHList

class HasResponseHeader h a headers where
  hlistLookupHeader :: HList headers -> ResponseHeader h a

instance {-# OVERLAPPING #-} HasResponseHeader h a (Header h a ': rest) where
  hlistLookupHeader (HCons ha _) = ha

instance {-# OVERLAPPABLE #-} (HasResponseHeader h a rest) => HasResponseHeader h a (first ': rest) where
  hlistLookupHeader (HCons _ hs) = hlistLookupHeader hs
```

Would a PR adding this be welcome? If so, where should it go?

**Comment by alpmestan** [comment]
Yes, that would be very welcome! It should go in [`Servant.API.ResponseHeaders`](https://github.com/haskell-servant/servant/blob/master/servant/src/Servant/API/ResponseHeaders.hs). Thanks in advance.

**Comment by phadej** [comment]
Very welcome indeed!

**Comment by phadej** [comment]
This is released in 0.15

**Comment by DanBurton** [comment]
Awesome, thanks!
